### PR TITLE
HPCC-25749 Create numDiskWrites statistic for logical files and index files

### PR DESCRIPTION
**Pull request by shamser**
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by shamser** [comment]
Tested by creating index and data file and then examining the file properties.  Separate jira to implement regression for new stats: https://track.hpccsystems.com/browse/HPCC-26008

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp [hunk 1]
```diff
@@ -243,6 +243,8 @@ class IndexWriteSlaveActivity : public ProcessSlaveActivity, public ILookAheadSt
             builder.clear();
             if (builderIFileIO)
             {
+                stat_type numDiskWrites = builderIFileIO->getStatistic(StNumDiskWrites);
+                stats.sumStatistic(StNumDiskWrites, numDiskWrites);
```
Comment (ghalliday) [inline thorlcr/activities/indexwrite/thindexwriteslave.cpp:247]: I think you want a more general:

                mergeStats(stats, builderIFileIO, diskWriteRemoteStatistics);

And a change to indexWriteActivityStatistics below.

File: thorlcr/thorutil/thormisc.cpp [hunk 2]
```diff
@@ -76,7 +76,7 @@ const StatisticsMapping basicActivityStatistics({StTimeLocalExecute, StTimeBlock
 const StatisticsMapping groupActivityStatistics({StNumGroups, StNumGroupMax}, basicActivityStatistics);
 const StatisticsMapping hashJoinActivityStatistics({StNumLeftRows, StNumRightRows}, basicActivityStatistics);
 const StatisticsMapping indexReadActivityStatistics({StNumRowsProcessed, StNumIndexSeeks, StNumIndexScans, StNumPostFiltered, StNumIndexWildSeeks}, basicActivityStatistics);
-const StatisticsMapping indexWriteActivityStatistics({StPerReplicated}, basicActivityStatistics);
+const StatisticsMapping indexWriteActivityStatistics({StPerReplicated, StNumDiskWrites}, basicActivityStatistics);
```
Comment (ghalliday) [inline thorlcr/thorutil/thormisc.cpp:79]: change to

    const StatisticsMapping indexWriteActivityStatistics({StPerReplicated}, basicActivityStatistics, diskWriteRemoteStatistics)

similar to the indexWriteActivityStats.

**Review by ghalliday — CHANGES_REQUESTED** [review]
@shamser a couple of suggests which will publish more useful stats in the activity, and simplify the code a little.

**Review by jakesmith — COMMENTED** [review]
apart from Gavin's comments, looks fine

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp [hunk 3]
```diff
@@ -243,6 +243,8 @@ class IndexWriteSlaveActivity : public ProcessSlaveActivity, public ILookAheadSt
             builder.clear();
             if (builderIFileIO)
             {
+                stat_type numDiskWrites = builderIFileIO->getStatistic(StNumDiskWrites);
+                stats.mergeStatistic(StNumDiskWrites, numDiskWrites);
```
Comment (ghalliday) [inline thorlcr/activities/indexwrite/thindexwriteslave.cpp:247]: You haven't changed this to the more general:

            mergeStats(stats, builderIFileIO, diskWriteRemoteStatistics);

Which should publish more than one disk write statistic (e.g., time, total written) to the activity stats.

**Review by ghalliday — CHANGES_REQUESTED** [review]

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp (hunk 3 above)
Comment (jakesmith) [reply thorlcr/activities/indexwrite/thindexwriteslave.cpp:247]: was about to say same (re. general mergeStats method)

**Review by jakesmith — CHANGES_REQUESTED** [review]
https://github.com/hpcc-systems/HPCC-Platform/pull/15030/files#r653428038

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp (hunk 3 above)
Comment (shamser) [reply thorlcr/activities/indexwrite/thindexwriteslave.cpp]: > https://github.com/hpcc-systems/HPCC-Platform/pull/15030/files#r653428038

Did you want me to change something in this file?

**Comment by shamser** [comment]
> https://github.com/hpcc-systems/HPCC-Platform/pull/15030/files#r653428038

@jakesmith I'm not sure what to do about this link.  It seems to point to thorlcr/activities/indexwrite/thindexwrite.cpp source file.

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp (hunk 3 above)
Comment (jakesmith) [reply thorlcr/activities/indexwrite/thindexwriteslave.cpp]: > You haven't changed this to the more general:
> 
> ```
>         mergeStats(stats, builderIFileIO, diskWriteRemoteStatistics);
> ```
> 
> Which should publish more than one disk write statistic (e.g., time, total written) to the activity stats.

@shamser - are you also going to publish some of the other stats. Gavin has suggested here?

**Comment by jakesmith** [comment]
> <Exception><Code>28</Code><Source>eclagent</Source><Message>System error: 28: SLAVE #2 [10.20.0.126:20120]: 28, CFile::open /var/lib/HPCCSystems/queries/mythor_20120/V2532622385_libW20210618-113043-4.so, Disk full : Failed to save dll: /var/lib/HPCCSystems/queries/mythor_20120/V2532622385_libW20210618-113043-4.so</Message></Exception>

@AttilaVamos - is it so tight of space?

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp (hunk 3 above)
Comment (shamser) [reply thorlcr/activities/indexwrite/thindexwriteslave.cpp]: @jakesmith I wasn't going to as I didn't need the additional stats for the cost calculation.  However, I've have done that now.

**Review by jakesmith — COMMENTED** [review]
Looks good to me.
Please squash

**Review by jakesmith — APPROVED** [review]
Should this be targeting 8.2.0 ?

File: thorlcr/activities/indexwrite/thindexwrite.cpp [hunk 4]
```diff
@@ -280,6 +280,10 @@ class IndexWriteActivityMaster : public CMasterActivity
                 bloom->setProp("@bloomProbability", pval.str());
             }
             container.queryTempHandler()->registerFile(fileName, container.queryOwner().queryGraphId(), 0, false, WUFileStandard, &clusters);
+            props.setPropInt64("@timeDiskWriteIO", statsCollection.getStatisticSum(StTimeDiskWriteIO));
```
Comment (ghalliday) [inline thorlcr/activities/indexwrite/thindexwrite.cpp:283]: I don't think you want all of these recorded in the file information - by using the merge function they will be available in the activity stats, and that should be sufficient.  @jakesmith what do you think is useful to store?

Comment (jakesmith) [reply thorlcr/activities/indexwrite/thindexwrite.cpp:283]: right, yes all will be at the activity level, but we want to propagate some to retain at a logical file level.

>Which should publish more than one disk write statistic (e.g., time, total written) to the activity stats.

@ghalliday - was you previous comment re. ensuring published (the index stats.) at activity level, but not propagating them to the logical file properties?

Reflecting on that (what's wanted at logical file level).
StSizeDiskWrite - should be very similar to what is already accounted for and published by the write activities in their @size and @compressedSize properties, so I don't think to add this.
StNumDiskRetries - I can't see anywhere that updates that statistic, so remove for now?

StTimeDiskWriteIO - not sure needed at logical file level, likely says more about the contention at the time, than anything else, although we could add for now, as cheap/free, and see if useful later.
NumDiskWrites - we do want this one.

Comment (ghalliday) [reply thorlcr/activities/indexwrite/thindexwrite.cpp:283]: Yes, I should have been clearer.

My comment from before was meant to be that if you use the mergeStatistic call then all the disk stats would automatically be published in the *activity* (fixing a bug in the process).  I would only publish to the file stats that aren't there already and are needed for the costs.  I think that is only the number of writes (as you were doing already).

**Review by ghalliday — APPROVED** [review]
@jakesmith please check you are happy with it.

**Review by jakesmith — APPROVED** [review]
@shamser - looks good, go ahead and squash

**Review by jakesmith — APPROVED** [review]

**Review by ghalliday — APPROVED** [review]

**Pull request merged**